lib_eio/path.ml

```diff
@@ -30,10 +31,22 @@ let with_lines path fn =
   let buf = Buf_read.of_flow flow ~max_size:max_int in
   fn (Buf_read.lines buf)
 
-let load path =
-  with_open_in path @@ fun flow ->
-  let buf = Buf_read.of_flow flow ~max_size:max_int in
-  Buf_read.take_all buf
+let load ((t:#Fs.dir), path) =
+  let st = t#stat path in
+  let mymax = min Sys.max_string_length Sys.max_array_length in
+  if Optint.Int63.(compare st.st_size (of_int mymax)) = 1 then
+    raise (Fs.File_too_large
+             (path, Invalid_argument "can't represent a string that long"));
+  with_open_in (t, path) @@ fun flow ->
+  let buf = Cstruct.create (Optint.Int63.to_int st.st_size) in
+  let rec loop buf got =
+    try
+      let n = Flow.read flow buf in
+      loop (Cstruct.shift buf n) (n + got)
+    with End_of_file -> got
```

Comment:
```suggestion
    match Flow.read flow buf with
    | got -> loop (Cstruct.shift buf n) (n + got)
    | exception End_of_file -> got
```
Otherwise, it's not tail-recursive.

Reply:
good catch ! I only looked at the call and ignored it was wrapped in a try catch.